op-bindings/etherscan/client.go

```diff
@@ -0,0 +1,196 @@
+		body, err := c.fetch(url)
+		if err != nil {
+			return rpcResponse{}, err
+		}
+
+		var response rpcResponse
+		err = json.Unmarshal(body, &response)
+		if err != nil {
+			return rpcResponse{}, fmt.Errorf("failed to unmarshal as rpcResponse: %w", err)
+		}
+		if response.Result == errRateLimited {
+			return rpcResponse{}, errors.New(strings.ToLower(errRateLimited))
+		}
+
+		return response, nil
+	})
+}
+
+func (c *client) FetchAbi(address string) (string, error) {
+	url := c.constructUrl(c.baseUrl, "getabi", address, "contract", fmt.Sprintf("address=%s", address))
+	response, err := c.fetchEtherscanApi(url)
+	if err != nil {
+		return "", err
+	}
+
+	abi, ok := response.Result.(string)
+	if !ok {
+		return "", fmt.Errorf("API response result is not expected ABI string")
+	}
+
+	return abi, nil
+}
+
+func (c *client) FetchDeployedBytecode(address string) (string, error) {
+	url := c.constructUrl(c.baseUrl, "eth_getCode", address, "proxy", fmt.Sprintf("address=%s", address))
+	response, err := c.fetchEtherscanRpc(url)
+	if err != nil {
+		return "", fmt.Errorf("error fetching deployed bytecode: %w", err)
+	}
+
+	bytecode, ok := response.Result.(string)
+	if !ok {
+		return "", errors.New("API response result is not expected bytecode string")
+	}
+
+	return bytecode, nil
+}
+
+func (c *client) FetchDeploymentTxHash(address string) (string, error) {
+	url := c.constructUrl(c.baseUrl, "getcontractcreation", address, "contract", fmt.Sprintf("contractaddresses=%s", address))
+	response, err := c.fetchEtherscanApi(url)
+	if err != nil {
+		return "", err
+	}
+
+	results, ok := response.Result.([]interface{})
+	if !ok {
+		return "", fmt.Errorf("failed to assert API response: %s is type of []txInfo", response)
+	}
+	if len(results) == 0 {
+		return "", fmt.Errorf("API response result is an empty array")
+	}
+
+	deploymentTxInfo, ok := results[0].(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("failed to assert API response result[0] is type of txInfo")
+	}
+
+	txHash, ok := deploymentTxInfo["txHash"].(string)
+	if !ok {
+		return "", fmt.Errorf("failed to assert API response result[0][\"txHash\"] is type of string")
+	}
+
+	return txHash, nil
+}
+
+func (c *client) FetchDeploymentTx(txHash string) (TxInfo, error) {
+	url := c.constructUrl(c.baseUrl, "eth_getTransactionByHash", txHash, "proxy", fmt.Sprintf("txHash=%s&tag=latest", txHash))
+	response, err := c.fetchEtherscanRpc(url)
+	if err != nil {
+		return TxInfo{}, err
+	}
+
+	resultBytes, err := json.Marshal(response.Result)
+	if err != nil {
+		return TxInfo{}, fmt.Errorf("failed to marshal Result into JSON: %w", err)
+	}
+
+	var tx TxInfo
+	err = json.Unmarshal(resultBytes, &tx)
+	if err != nil {
+		return TxInfo{}, fmt.Errorf("API response result is not expected txInfo struct: %w", err)
+	}
+
+	return tx, nil
+}
+
+func (c *client) constructUrl(baseUrl, action, address, module, params string) string {
```

Comment:
```suggestion
func constructUrl(baseUrl, action, address, module, params string) string {
```

ffti: no need for this to be a method on client since it's just a pure function based on the inputs